### PR TITLE
Support remote for Python UDF in distributed autograd

### PR DESCRIPTION
**Pull request by mrshenli**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#28656 Support remote for Python UDF in distributed autograd**
* #28630 Support remote for builtin operators in distributed autograd

Differential Revision: [D18138561](https://our.internmc.facebook.com/intern/diff/D18138561)

### CODE REVIEW

**Comment by zhaojuanmao** [comment]
overall lgtm, same comments as #28630

File: test/dist_autograd_test.py [hunk 1]
```diff
@@ -433,10 +501,24 @@ def test_no_graph_with_tensors_not_require_grad(self):
...
+                # the callee, as the caller does not know whether the return
+                # value would contain a requires_grad tensor or not.
+                pass
+
+    @dist_init(setup_model_parallel=True)
+    def test_no_graph_with_tensors_not_require_grad(self):
+        self._test_no_graph_with_tensors_not_require_grad(ExecMode.RPC_SYNC)
 
+    @dist_init(setup_model_parallel=True)
+    def test_no_graph_with_tensors_not_require_grad_remote(self):
```
Comment (pritamdamania87) [inline test/dist_autograd_test.py]: Can we test nested RPCs with RRef? For example A shares RRef to B, B shares it to C, C calls to_here() does some computation and finally the result is retrieved on A and A calls backward on this. (The RRef could live on D).

File: torch/csrc/distributed/rpc/rref.cpp [hunk 2]
```diff
@@ -174,14 +174,36 @@ template <>
...
+          rpcWithAutograd.tensors(),
+          rpcWithAutograd.fromWorkerId());
+
+      auto& wrappedRpc = rpcWithAutograd.wrappedRpc();
+      auto& rfr = static_cast<PythonRRefFetchRet&>(wrappedRpc);
+      future->markCompleted(c10::ivalue::Tuple::create(rfr.values()));
+    } else {
+      auto rfr = PythonRRefFetchRet::fromMessage(message);
+      future->markCompleted(c10::ivalue::Tuple::create(rfr->values()));
+    }
```
Comment (pritamdamania87) [inline torch/csrc/distributed/rpc/rref.cpp]: Can we dedup this code? I think we had something similar in ScriptRRefFetchRet.

**Comment by mrshenli** [comment]
Created https://github.com/pytorch/pytorch/issues/28819 to track the coverage on a corner case.

File: test/dist_autograd_test.py [hunk 3]
```diff
@@ -688,6 +732,139 @@ def test_backward_simple(self):
...
+        )
+
+        # kick off forward and backward pass on three other workers (trainers)
+        rank_diffs = [1, 2, 3]
+        futures = []
+        for rank_diff in rank_diffs:
+            futures.append(rpc.rpc_async(
+                "worker{}".format((self.rank + rank_diff) % self.world_size),
+                _run_trainer,
+                args=(rref_t1, t2, t2.grad, self_name, rank_diff)
```
Comment (pritamdamania87) [inline test/dist_autograd_test.py]: Not sure I follow the test completely. It looks like the `ps` is different for each node (the ps is self.rank). Shouldn't make rank 0 the ps and ranks 1, 2 and 3 be trainers?

File: test/dist_autograd_test.py [hunk 4]
```diff
@@ -688,6 +732,139 @@ def test_backward_simple(self):
...
+                "torch::distributed::autograd::RecvRpcBackward",
+                next_funcs[1][0].name()
+            )
+
+        # unblock trainers
+        _set_rpc_done(None, 0)
+
+        # wait until all trainers are done
+        for fut in futures:
+            fut.wait()
```
Comment (pritamdamania87) [inline test/dist_autograd_test.py]: We should validate that grads are accumulated appropriately on the PS.

File: torch/csrc/distributed/autograd/engine/dist_engine.cpp [hunk 5]
```diff
@@ -246,6 +246,7 @@ void DistEngine::execute(const variable_list& roots) {
   std::shared_ptr<Node> graphRoot =
       std::make_shared<GraphRoot>(rootEdges, grads);
   edge_list outputEdges;
+
```
Comment (pritamdamania87) [inline torch/csrc/distributed/autograd/engine/dist_engine.cpp]: nit: extra empty line?

File: torch/csrc/distributed/rpc/python_functions.cpp [hunk 6]
```diff
@@ -30,14 +30,14 @@ std::shared_ptr<Operator> matchBuiltinOp(
     const py::kwargs& kwargs,
     Stack& stack) {
   Symbol symbol = Symbol::fromQualString(opName);
+
```
Comment (pritamdamania87) [inline torch/csrc/distributed/rpc/python_functions.cpp]: nit: extra empty line?

File: torch/csrc/distributed/rpc/python_functions.cpp [hunk 7]
```diff
@@ -163,6 +163,7 @@ std::shared_ptr<FutureMessage> pyRpcPythonUdf(
   auto pythonUDFCall = c10::guts::make_unique<PythonUDFCall>(
       std::vector<char>(pickledPythonUDF.begin(), pickledPythonUDF.end()),
       tensors);
+
```
Comment (pritamdamania87) [inline torch/csrc/distributed/rpc/python_functions.cpp]: nit: extra empty line?

File: test/dist_autograd_test.py (hunk 3 above)
Comment (mrshenli) [reply test/dist_autograd_test.py]: This is testing 4 ps and 12 trainers concurrently. Each process x will serve as a ps and driver which will kick off 3 trainers that only talks to ps x. So we have

ps = rank0, trainer = rank1, 2, 3
ps = rank1, trainer = rank 2, 3, 4
....

Let me add some comments here to explain.

File: test/dist_autograd_test.py [hunk 8]
```diff
@@ -688,6 +732,139 @@ def test_backward_simple(self):
...
+            self.assertEqual(2, len(next_funcs))
+            self.assertEqual(
+                "torch::autograd::AccumulateGrad", next_funcs[0][0].name()
+            )
+            if accumulate_grad_func:
+                self.assertEqual(accumulate_grad_func, next_funcs[0][0])
+            else:
+                accumulate_grad_func = next_funcs[0][0]
+            self.assertEqual(
+                "torch::distributed::autograd::RecvRpcBackward",
```
[the remark this answers is not in the document]
Comment (mrshenli) [reply test/dist_autograd_test.py]: good point, let me make the change.

File: test/dist_autograd_test.py [hunk 9]
```diff
@@ -69,6 +105,17 @@ def _all_contexts_cleaned_up(timeout_seconds=10):
     return success
 
 
+# This function creates a dis atugorad context, run rpc_sync on the given ps,
+# and then blocks until the ps has verified the grads are correctly accumulated.
+def _run_trainer(rref_t1, t2, t2_grad, ps, rank_diff):
```
Comment (pritamdamania87) [inline test/dist_autograd_test.py]: t2_grad is not used here?

**Review by pritamdamania87 — APPROVED** [review]

**Pull request closed without merging**